### ISSUE TITLE
Sandboxes can't be updated concurrently

### ISSUE THREAD

**Issue by jcpetruzza** [haskell/cabal#2220]
I am sharing a sandbox between several related projects and have an initialization script that:

 1) sets up the main sandbox (e.g. `cd global-sandbox; cabal sandbox init`)
 2) adds some sources (e.g., `cabal sandbox add-source patched-library`)
 3) for each project, initializes the shared sandbox (e.g. `cd project; cabal sandbox init --sandbox=global-sandbox`).

This setup worked fine until I tried to run step 3 in parallel. Now cabal suddenly complains that `global-sandbox/add-source-timestamps.tmp does not exist`.

Looking at Distribution/Client/Sandbox/Timestamp.hs, there seems to be indeed a race condition in `writeTimestampFile`, where one process tries to rename a file that was already renamed by another process.

**Comment by 3noch** [comment]
I think there might be a worse race-condition in 1.22. Simply installing dependencies randomly fails with `permission denied`. It appears it's trying to open a file that is locked. It happens randomly. I'm on Windows.

**Comment by acowley** [comment]
I hit this error, too. My Nix tooling produces a build plan for each upstream dependency using the downstream sandbox to pick up add-source'ed dependencies. I added the capability to produce these build plans in batches of parallel jobs, but they fight over this timestamps tmp file. I'm not sure that these processes actually have to modify the timestamps file in this case, so perhaps they could just skip the bit that tries to update the file?

**Comment by 3noch** [comment]
I have had good success with the `filelock` package. It has native support for Windows and unix and has system level blocking on the lock (i.e. no `threadDelay`s, thus very responsive). On Windows at least, it automatically unlocks the lock on process death.

**Comment by BardurArantsson** [comment]
Is "don't do that" an acceptable answer? I think we'll instead want to address the overall use case instead of diving into low-level details such as file locking (over NFS? Shudder). What is the use case that is behind this? Why isn't cabal-install doing that for you? Give us meta-goals! :)

**Comment by 3noch** [comment]
One advantage is if you're using `cabal` as part of a CI system using a shared sandbox. I don't know if two concurrent calls to `cabal update` would cause problems, but certainly `cabal install --dependencies-only` with a shared sandbox would be problematic. With stackage, it's pretty common to have widely used sandboxes.

**Comment by BardurArantsson** [comment]
Are you doing some sort of "cabal install ..." thing with two concurrent CI instances? (I think that's what I can gather from your comment).

I think that's actually a use case! :)

**Comment by 3noch** [comment]
Yes. I use Jenkins CI and my projects call `cabal update` and `cabal install --dependencies-only` using a shared sandbox (thanks to `stackage sandbox init`). You're right that it's possible to handle the locking at a higher level, but it's a bit surprising that `cabal` does not support concurrent behavior in cases when things are explicitly shared.

**Comment by 3noch** [comment]
In other words, it would simply make `cabal` much friendlier to these use cases if it supported this. As it is, I'd have to write my own file-locking mechanism which would have to be aware of what's being shared (for example, different stackage sandboxes need not share the same lock). That becomes complex fast.

**Comment by BardurArantsson** [comment]
That's absolutely fair enough, thank you for a good use case. Jenkins CI is already pretty underpowered (for Free projects, at least), so anything that cabal/cabal-install could do in that setting  would be valuable.

Btw, I apologize if I seemed blunt or aggressive... it that many times I've gone asking people if they *actually* need what they're asking for (or their actual *use case* vs. what they think they need). Often these sets turn out different.

**Comment by 3noch** [comment]
@BardurArantsson I might only be speaking for myself, but no offense taken whatsoever. You're right that have at least one *specific* use-case is helpful, if only for testing.

**Comment by acowley** [comment]
My use case is what I wrote previously. It's a shared sandbox that a bunch of build plans are computed against.

I do accept that the de facto answer is "don't do that," so I'm working to limit usage of `cabal-install`.

**Comment by BardurArantsson** [comment]
@acowley I apologize... I didn't notice you in the noise :). I take it that the Nix tooling also tries to parallelize pretty aggressively? (As it would, being "pure" an' all.)

**Comment by BardurArantsson** [comment]
(I should say: I have no official "clout" with the Cabal project. I was just going through a lot of issues and happened to comment on "yours".)

**Comment by acowley** [comment]
@BardurArantsson No problem! I had tried to parallelize one part of things I was doing because it seemed entirely reasonable, but ran into this issue, so I chimed in above. It's not a huge deal for me, but I do think the existing approach is slightly too pessimistic about locking things down.

I can get around this by invoking `cabal` once for multiple projects, then parsing and splitting the resulting build plan and just doing more of the building work myself.
